openreader/src/db.ts

```diff
@@ -1,83 +1,72 @@
-import type {ClientBase, Pool, PoolClient, PoolConfig} from "pg"
+import type {ClientBase, Pool} from "pg"
+import {OpenreaderContext} from "./context"
+import {Dialect} from "./dialect"
+import {Query} from "./sql/query"
+import {Subscription} from "./subscription"
+import {withErrorContext} from "./util/error-handling"
+import {LazyTransaction} from "./util/lazy-transaction"
 
 
 export interface Database {
     query(sql: string, parameters?: any[]): Promise<any[][]>
-    escapeIdentifier(name: string): string
-}
-
-
-/**
- * This is an interface OpenReader uses to interact with underling database.
- */
-export interface Transaction {
-    get(): Promise<Database>
 }
 
 
 export class PgDatabase implements Database {
     constructor(private client: ClientBase) {}
 
-    query(sql: string, parameters?: any[]): Promise<any[]> {
-        return this.client.query({text: sql, rowMode: 'array'}, parameters).then(result => result.rows)
-    }
-
-    escapeIdentifier(name: string): string {
-        return this.client.escapeIdentifier(name)
+    query(sql: string, parameters?: any[]): Promise<any[][]> {
+        return this.client.query({text: sql, rowMode: 'array'}, parameters)
+            .then(result => result.rows)
+            .catch(withErrorContext({sql, parameters}))
     }
 }
 
 
-export class PoolTransaction implements Transaction {
-    private tx: Promise<{client: PoolClient, db: Database}> | undefined
-    private closed = false
+export class PoolOpenreaderContext implements OpenreaderContext {
+    private tx: LazyTransaction<Database>
+    private subscriptionPool: Pool
 
-    constructor(private pool: Pool) {}
+    constructor(
+        public readonly dialect: Dialect,
+        pool: Pool,
+        subscriptionPool?: Pool,
+        private subscriptionPollInterval: number = 1000
```

Comment:
It's better to set poll interval to 5 seconds since the average block time is about 6 seconds.